### PR TITLE
Update input value only after value applying for SelectBox with allowed custom items (T851874)

### PR DESCRIPTION
**Pull request by alexander-kotov-dx**

### CODE REVIEW

File: testing/tests/DevExpress.ui.widgets.editors/selectBox.tests.js [hunk 1]
```diff
@@ -1740,6 +1740,29 @@ QUnit.module('editing', moduleSetup, () => {
...
+            searchEnabled: true,
+            searchTimeout: 0
+        });
+        const $input = $element.find(toSelector(TEXTEDITOR_INPUT_CLASS));
+        const keyboard = keyboardMock($input);
+        const instance = $element.dxSelectBox('instance');
+        const $ddButton = $element.find(toSelector(DX_DROP_DOWN_BUTTON));
+
+        $input.get(0).setSelectionRange(0, 2);
+        keyboard.caret({ start: 0, end: 2 });
```
Comment (DokaRus) [inline testing/tests/DevExpress.ui.widgets.editors/selectBox.tests.js]:
```suggestion
        keyboard
            .caret({ start: 0, end: 2 })
            .press('backspace');
```

File: testing/tests/DevExpress.ui.widgets.editors/selectBox.tests.js [hunk 2]
```diff
@@ -1740,6 +1740,29 @@ QUnit.module('editing', moduleSetup, () => {
...
+        keyboard.press('backspace');
```
Comment (DokaRus) [inline testing/tests/DevExpress.ui.widgets.editors/selectBox.tests.js]:
```suggestion

```

File: testing/tests/DevExpress.ui.widgets.editors/selectBox.tests.js [hunk 3]
```diff
@@ -1740,6 +1740,29 @@ QUnit.module('editing', moduleSetup, () => {
...
+            acceptCustomValue: true,
+            searchEnabled: true,
+            searchTimeout: 0
+        });
+        const $input = $element.find(toSelector(TEXTEDITOR_INPUT_CLASS));
+        const keyboard = keyboardMock($input);
+        const instance = $element.dxSelectBox('instance');
+        const $ddButton = $element.find(toSelector(DX_DROP_DOWN_BUTTON));
+
+        $input.get(0).setSelectionRange(0, 2);
```
Comment (DokaRus) [inline testing/tests/DevExpress.ui.widgets.editors/selectBox.tests.js]: Is it didn't help?
```
keyboardMock($input, true)
<...>
keyboard.caret({ start: 0, end: 2 })
```

**Review by DokaRus — DISMISSED** [review]

File: js/ui/select_box.js [hunk 4]
```diff
@@ -501,7 +501,7 @@ const SelectBox = DropDownList.inherit({
         this.callBase();
     },
 
-    _restoreInputText: function() {
+    _restoreInputText: function(saveEditingValue) {
```
Comment (San4es) [inline js/ui/select_box.js:504]: Do we need to take into account the `saveEditingValue` argument on the TagBox side?

File: js/ui/select_box.js [hunk 5]
```diff
@@ -510,7 +510,10 @@ const SelectBox = DropDownList.inherit({
             const initialSelectedItem = this.option('selectedItem');
 
             if(this.option('acceptCustomValue')) {
```
Comment (San4es) [inline js/ui/select_box.js:512]:
```suggestion
            if(this.option('acceptCustomValue') && !saveEditingValue) {
                this._updateField(initialSelectedItem);
                this._clearFilter();
                return;
           }
```

File: js/ui/select_box.js (hunk 4 above)
Comment (alexander-kotov-dx) [reply js/ui/select_box.js:504]: Thanks, it is a good idea. The TagBox does not hat have the initial issue, but we can unify its behavior with dxSelectBox

File: js/ui/select_box.js (hunk 5 above)
Comment (alexander-kotov-dx) [reply js/ui/select_box.js:512]: i think we should not change behavior for searchMode without acceptCustomValue in this case

**Review by San4es — APPROVED** [review]
LGTM

**Review by DokaRus — APPROVED** [review]

**Pull request merged**